[EuiTableHeader] Add prop to remove `<tr>`

K-Kumar-01 opened the pull request:
### Summary

Issue number #4183 
Added the ability to remove the `<tr/>` tag.

### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [x] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#adding-playground-toggles)**
- ~[ ] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**~
- ~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples~
- [x] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

cchaos commented on src/components/table/table_header.tsx:
```diff
@@ -20,14 +20,23 @@
 import { CommonProps } from '../common';
 
-export const EuiTableHeader: FunctionComponent<CommonProps> = ({
+type Props = CommonProps & {
+  withInbuiltTag?: boolean;
+};
```
I think we can name this prop a little better and add a prop comment to fill out the props table. Also, since there's more than just `CommonProps`, we'll want to export the TS type by also naming it more specific to this component.

```suggestion
export type EuiTableHeaderProps = CommonProps & {
  /**
   * Automatically adds a wrapping `<tr>` element around the children
   */
  wrapWithTableRow?: boolean;
};
```

cchaos replied:
Then you'll need to update the rest of the file to match

cchaos commented on src/components/table/table_header.tsx:
```diff
@@ -20,14 +20,23 @@
-      <tr>{children}</tr>
+      {withInbuiltTag ? (
+        <tr>{children}</tr>
+      ) : (
+        <React.Fragment>{children}</React.Fragment>
+      )}
```
I don't think the Fragment is necessary here. You should just be able to do:
```suggestion
      ) : (
        {children}
      )}
```

cchaos reviewed (changes requested):
Thanks for tackling this one @K-Kumar-01 . There are still some more necessary things to complete this PR,  I'll update the checklist in PR summary by crossing out the ones that aren't necessary. But at top level, we'll at least need a new test for this prop and a changelog entry.

cchaos commented:
As a side-note, when creating new PR's it's always best to start a new branch from the latest `master` branch. That way, your commit log will only include commits directly related to this PR.

K-Kumar-01 commented:
> As a side-note, when creating new PR's it's always best to start a new branch from the latest `master` branch. That way, your commit log will only include commits directly related to this PR.

I will keep in mind the next time I make a PR.:)

K-Kumar-01 commented:
@cchaos 
I have made the required changes in the file.
However, I am facing issues with writing tests as I have not written any tests earlier.
![Screenshot from 2021-02-02 12-04-35](https://user-images.githubusercontent.com/59891164/106561526-d88e7000-654e-11eb-97da-f972e58491ba.png)
I am not able to understand how to rectify this one. In `components/basic_table/collapsed_items_actions.tsx` also has a similar approach. However, my one gives error

cchaos commented:
@K-Kumar-01 Can you commit what you currently have? I'm guessing it has to do with the actual `render()` of the component without the `<tr>` and not a fault of the test. (The test doing the right thing 😉 )

K-Kumar-01 commented:
@cchaos I have committed the recent changes. Let me know if there are any changes requried.

cchaos reviewed (approved):
So it looks like the way our testing suite is managed, is there is some level of DOM validation happening when creating snapshots. So that meant, when you were providing a simple `string` as the child of a `<tr>`, it said "nope 😆 " and appended the invalid child outside of the `<tr>` instead. So I fixed up the tests by passing the valid DOM structure and it correctly renders the snapshots.

I also updated the name of the prop.

👍  LGTM

cchaos commented:
Jenkins, test this

cchaos commented:
Jenkins, test this

K-Kumar-01 commented:
> So it looks like the way our testing suite is managed, is there is some level of DOM validation happening when creating snapshots. So that meant, when you were providing a simple `string` as the child of a `<tr>`, it said "nope laughing " and appended the invalid child outside of the `<tr>` instead. So I fixed up the tests by passing the valid DOM structure and it correctly renders the snapshots.
> 
> I also updated the name of the prop.
> 
> +1 LGTM

Thanks a lot for the tests. I will try my best to write learn writing tests :smiley: and also ensure to pass Valid DOM structure :grimacing:

cchaos commented:
Oh no worries. It caught it me off guard too and have to play around with the `children` I passed to see exactly what was happening. First time I've seen this, so good find! 😜 

Jenkins, test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4465/

cchaos commented:
Jenkins, test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4465/
